test/test_allocation_domain.cpp

```diff
@@ -303,4 +303,306 @@ TEST_F(AllocationDomainTest, Tensor3d_To_NHWC4d_FwdBwd_CUDA) {
+
+  testValidate(&fusion, cg_outputs, {t0}, {t0}, __LINE__, __FILE__);
+}
+
+// A global->global copy kernel where both inputs are NHWC memory format. The
+// allocation domain view both the input and the output as a 1d tensors.
+TEST_F(AllocationDomainTest, NHWC1d_To_NHWC1d_CUDA) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  int n = 31, h = 64, w = 103, c = 21;
+
+  auto tv0 = makeContigConcreteTensor({n * h * w * c});
+  fusion.addInput(tv0);
+
+  std::vector<IterDomain*> tv0_1d = {tv0->axis(0)};
+  tv0->setAllocationDomain(tv0_1d, true);
+  tv0->split(0, c);
+  tv0->split(0, w);
+  tv0->split(0, h);
+  tv0->reorder({{-1, 1}});
+  tv0->commitLeafToRFactor();
+
+  auto tv1 = set(tv0);
+  fusion.addOutput(tv1);
+
+  // [N, C, H, W] -> [N, H, W, C]
+  tv1->reorder({{1, -1}});
+  // [N, H, W, C] -> [N*H*W*C]
+  tv1->merge(0);
+  tv1->merge(1);
+  tv1->merge(0);
+
+  std::vector<IterDomain*> tv1_1d = {tv1->axis(0)};
+  tv1->setAllocationDomain(tv1_1d, true);
+
+  tv1->split(0, 4);
+  tv1->axis(1)->parallelize(ParallelType::Vectorize);
+  tv1->split(0, 128);
+  tv1->axis(1)->parallelize(ParallelType::TIDx);
+  tv1->axis(0)->parallelize(ParallelType::BIDx);
+  // [BIDx, TIDx, V]
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+  at::Tensor t0_wrong_format = at::randn({n, c, h, w}, options);
+  at::Tensor t0 = t0_wrong_format.contiguous(at::MemoryFormat::ChannelsLast);
+
+  FusionExecutor fe;
+  fe.compileFusion(fusion_ptr.get(), {t0});
+
+  EXPECT_THAT(
+      [&]() { fe.runFusion({t0_wrong_format}); },
+      ::testing::ThrowsMessage<c10::Error>(::testing::HasSubstr(
+          "Splitting one dimension into discontiguous dimensions is not")));
+
+  auto cg_outputs = fe.runFusion({t0});
+
+  ASSERT_TRUE(cg_outputs[0].is_contiguous(at::MemoryFormat::ChannelsLast));
+
+  testValidate(&fusion, cg_outputs, {t0}, {t0}, __LINE__, __FILE__);
+}
+
+// A global->global copy kernel where both inputs are NHWC memory format. The
+// allocation domain view the input as a 2d tensor of shape [N*H/8, 8*W*C], and
+// view the output as a 2d tensor of shape [N*H*W*C/4, 4]
+TEST_F(AllocationDomainTest, NHWC2d_To_NHWC2d_CUDA) {
+  auto fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  int n = 31, h = 64, w = 103, c = 21;
+
+  auto tv0 = makeContigConcreteTensor({n * h / 8, 8 * w * c});
+  fusion.addInput(tv0);
+
+  std::vector<IterDomain*> tv0_2d = {tv0->axis(0), tv0->axis(1)};
+  tv0->setAllocationDomain(tv0_2d, true);
+  tv0->merge(0);
+  tv0->split(0, c);
+  tv0->split(0, w);
+  tv0->split(0, h);
+  // [N, H, W, C]
+  tv0->reorder({{-1, 1}});
+  // [N, C, H, W]
+  tv0->commitLeafToRFactor();
+
+  auto tv1 = set(tv0);
+  fusion.addOutput(tv1);
+
+  std::vector<IterDomain*> tv0_nhwc = {
+      tv0->axis(0), tv0->axis(2), tv0->axis(3), tv0->axis(1)};
+  tv0->setAllocationDomain(tv0_nhwc, true);
```

Comment:
you are losing me here...

Wasn't the NHWC definition already done at the `tv0->commitLeafToRFactor()` above? what's the setAllocationDomain on tv0 doing again?!

Comparing this test with `NHWC1d_To_NHWC4d_CUDA` above, why's `setAllocationDomain` needed here?

Reply:
I realize that this is likely my lack of understanding on the relationship between leaf domain / rfactor domain / allocation domain. So is there a good doc somewhere that roughly summarizes those?

Reply:
Unfortunately I don't think we have doc for them. We don't have a doc describing what is rFactor domain, do we? If we do have a comment in the code, I would be happy to add allocation domain to that comment.

I suggest to just think of `commitLeafToRFactor` as a convenient method for defining a tensor. For example, if we want to define a tensor
```
root: [I1, I2, I3]
rFactor: [I3, I1*I2]
```
We can of course define it manually by creating all the `IterDomain*`, `Merge*` for the root and rFactor domain, then manually create `TensorDomain*` and `TensorView*`, but this is not convenient. It is more convenient to just use our already-built utilities like `TensorView::merge`, `TensorView::reorder`, and the `commitLeafToRFactor` is exactly for that purpose.

`commitLeafToRFactor` has nothing to do with memory format, and calling `commitLeafToRFactor` means nothing about memory format. Defining the rFactor domain of a tensor using `commitLeafToRFactor` is nothing different from the plain old way of defining a tensor with rFactor domain. This method only start being heavily used in allocation domain PRs does not mean it is a utility for allocation domain. It is just because this utility is super new, and it has not been used widely yet.

The recommended mindset for reasoning about `commitLeafToRFactor` should not be "`commitLeafToRFactor` changes the rFactor domain from XXX to XXX", because this would make it confusing. The mindset I am expecting people to read codes with `commitLeafToRFactor` is, when you see code like:
```C++
auto tv = makeContigTensor(4);
tv->....;
tv->....;
tv->....;
tv->commitLeafToRFactor();
```
you will know that the code block between creating `tv` and `commitLeafToRFactor` are the region that you define this tensor. So you should automatically add some comment in your mind
```C++
auto tv = makeContigTensor(4);  // begin defining tv. From now on, I will build a graph of IterDomains connecting root and rFactor domain
tv->....;
tv->....;
tv->....;
tv->commitLeafToRFactor();  // end defining tv. OK, the rFactor domain is built. The relationship between root and rFactor domain is blabla, so this is a view+permute op that changes the tensor semantics from XXX to YYY.
```

Reply:
Whether a tensor is NHWC or NCHW depend on whether the order of `IterDomain`s of rFactor domain and allocation domain match. The key is whether rFactor match with allocation, and it has nothing to do with the relationship between rFactor domain and root domain.

Reply:
The rFactor domain should always be `NCHW`, regardless of the memory format, and if allocation domain match with rFactor domain, then this tensor's memory format is also `NCHW`. But if the order of allocation domain is `[N, H, W, C]`, or more complicated `[N*H*W*C]`, `[N*H, W*C]`, `[N, H*W, C]`, `[N, H, W, C/5, 5]`, then this tensor a NHWC tensor.

Reply:
Also, rFactor domain only has something to do with tensor semantics, it has nothing to do with memory format. This is just like in PyTorch, having a tensor with semantics `NCHW` does not means it is a `NCHW` tensor, and doing a permute on that tensor to change its semantics `NHWC` does not mean you changed the memory format from `NCHW` to `NHWC`.

Reply:
Thanks a lot. This sounds like what I remembered in our previous conversations, but having it written down here is helpful.

I didn't realize that we can have view/permute setup directly on input tensor, but I guess that makes sense.

One confusion is still bothering me even after the read.
```
 std::vector<IterDomain*> tv0_2d = {tv0->axis(0), tv0->axis(1)};
  tv0->setAllocationDomain(tv0_2d, true);
  tv0->merge(0);
  tv0->split(0, c);
  tv0->split(0, w);
  tv0->split(0, h);
  // [N, H, W, C]
  tv0->reorder({{-1, 1}});
  // [N, C, H, W]
  tv0->commitLeafToRFactor();

  auto tv1 = set(tv0);
  fusion.addOutput(tv1);

  std::vector<IterDomain*> tv0_nhwc = {
      tv0->axis(0), tv0->axis(2), tv0->axis(3), tv0->axis(1)};
  tv0->setAllocationDomain(tv0_nhwc, true);
```
What's the purpose of `tv0->setAllocationDomain(tv0_2d, true);`?
if all transformation done until `commitLeafToRFactor()` is only semantic-related, that allocation assign in the beginning doesn't seem to be doing anything.
And later we are setting allocation domain explicitly again. which should overwrite the first allocation setup?!

Reply:
Oh, I see where the confusion is. I copy-pasted tests from my previous prototyping branch. I remembered that these tests were already well written and cleaned up. Seems that my memory was wrong. Sorry about the confusion. I will update the test. And yes, your understanding is correct. Setting the allocation domain twice overwrite the first result, so it makes no sense to do so.

Reply:
Updated the test

Reply:
Updated the test